software/py/bsg_manycore_link_gen.py

```diff
@@ -0,0 +1,252 @@
+  
+  Manycore's PC width is 24-bit, addressing a total of 16MB program space.
+  Since the 31:25 MBSs are assumed to be 0s, linking should also assume
+  0x0-0x01000000 as the valid program space. This means that .text section VMAs
+  lie in 0x0-0x01000000. On the other hand, .text section is loaded to DRAM and
+  as result, it's LMAs shouls be >0x80000000. So, .text section is loaded in the
+  region 0x80000000-0x81000000. Consequently, VMA's 0x80000000-0x81000000 are
+  are reserved for .text section as well.
+  """
+
+  _opening_comment = \
+    "/**********************************\n" \
+    + " BSG Manycore Linker Script \n\n"
+
+  def __init__(self, default_data_loc, dram_size, sp):
+    self._default_data_loc = default_data_loc
+    self._dram_size = dram_size
+    self._sp = sp
+    self._opening_comment += \
+        " data default: {0}\n".format(default_data_loc) \
+        + " dram memory size: 0x{0:08x}\n".format(dram_size) \
+      + "***********************************/\n"
+
+  # Format:
+  # <mem region name> : [<access>, <start address>, <size>]
+  def _memory_regions(self, mem_regions):
+    script = 'MEMORY {\n'
+
+    for m, access, origin, size in mem_regions:
+      script += "{0} ({1}) : ORIGIN = 0x{2:08x}, LENGTH = 0x{3:08x}\n".format(
+          m, access, origin, size)
+
+    script += '}\n'
+
+    return script
+
+  def _section(self, name, address, vma_region, lma_region,
+      in_sections, in_objects):
+    """
+    Forms an LD section.
+    
+    Arguments:
+    name         -- name of the section.
+    address      -- if `vma_region` is null, `address` is assumed to
+                    be section's logical address. If `lma_region` is null,
+                    `address` is assumed to be sections load address.
+    vma_region   -- name of the VMA region if the section is to be directed to
+                    a VMA region, else None.
+    lma_region   -- name of the LMA region if the section is to be directed to
+                    a VMA region, else None.
+    in_sections  -- list of input sections.
+    in_objects   -- list of input objects to consider.
+    """
+    vaddr        = "" # virtual address
+    laddr        = "" # logical address
+    vma_redirect = ""
+    lma_redirect = ""
+
+    if lma_region == None:
+      laddr = "AT({0})".format(address)
+      vma_redirect = ">{0}".format(vma_region)
+    else:
+      vaddr = address
+      lma_redirect = "AT>{0}".format(lma_region)
+
+    script = "{0} {1} :\n{2} {{".format(name, vaddr, laddr)
+
+    for sec in in_sections:
+      script += "\n  {0}({1})".format(in_objects, sec)
+
+    script += "\n  . = ALIGN(8);\n"
+    script += "\n}} {0} {1}\n\n".format(vma_redirect, lma_redirect)
+
+    return script
+
+  def script(self):
+    # Address constants:
+    #
+    # LMA (Load Memory Address)    => NPA used by loader
+    # VMA (Virtual Memory Address) => Logical address used by linker for 
+    #                                 symbol resolutions
```

Comment:
I like the """ comments for blocks better than the #

Reply:
I think docstrings are generally better because they can be accessed with help() and I think there are tools to make documentation pages from them.

Reply:
Yes, that's why I like them. 

But you generally have to be pretty pedantic with formats.

Reply:
This is not """ because it's not the description of the method, it's just the description of statements below.

Anyway, I realized a description to this method makes sense explaining the VMA and LMA regions. So added it.